### ISSUE TITLE
bug: Starship fails to build through home-manager.

### ISSUE THREAD

**Issue by Icy-Thought** [nix-community/home-manager#2271]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

Enabling starship through home-manager w/o configurations (`programs.starship.enable = true;`) returns the following error:
```
/dev/fd/63: line 21: STARSHIP_PREEXEC_READY: unbound variable
```

Starship works as expected when installed through Nix though.

### Maintainer CC

@marsam 

### System information

```markdown
- system: `"x86_64-linux"`
 - host os: `Linux 5.13.9-zen1, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210802_47e96bb`
 - channels(sirius): `""`
 - channels(root): `"nixos-21.11pre306621.439b1605227"`
 - nixpkgs: `/etc/nixpkgs`
```

**Comment by rycee** [comment]
Hmm, haven't seen this before. If it happens at starship build time then there isn't much we can do about it, though. The build happens entirely on the Nixpkgs side. Dunno why it would work with plain Nix, perhaps you you are using different versions of Nixpkgs in your HM configuration and globally?

**Comment by lilyball** [comment]
This sounds like an issue with starship itself. If you run `starship init bash --print-full-init` you'll find that the error occurs in this function:

```bash
  15 # Will be run before *every* command (even ones in pipes!)
  16 starship_preexec() {
  17     # Save previous command's last argument, otherwise it will be set to "starship_preexec"
  18     local PREV_LAST_ARG=$1
  19 
  20     # Avoid restarting the timer for commands in the same pipeline
  21     if [ "$STARSHIP_PREEXEC_READY" = "true" ]; then
  22         STARSHIP_PREEXEC_READY=false
  23         STARSHIP_START_TIME=$(/nix/store/l3ppjhqlg6fnlyd2wgq205xsp3xz7ph3-home-manager-path/bin/starship time)
  24     fi
  25 
  26     : "$PREV_LAST_ARG"
  27 }
```

See how line 21 references the variable `$STARSHIP_PREEXEC_READY`. This function is installed with `trap 'starship_preexec "$_"' DEBUG`, so it runs before every command (run the aforementioned command if you want to see more comments about this).

The variable `$STARSHIP_PREEXEC_READY` is set in another function:

```bash
  29 # Will be run before the prompt is drawn
  30 starship_precmd() {
  ⋮   
  57     STARSHIP_PREEXEC_READY=true  # Signal that we can safely restart the timer
  58 }
```

This is invoked prior to each prompt by setting `PROMPT_COMMAND="starship_precmd"`.

The net result of this is `starship_preexec` is going to be invoked prior to `starship_precmd`.

And so, what I think is going on here, is that you have `set -u` (or `set -o nounset`) defined in your `~/.profile` or `~/.bash_profile` or `~/.bashrc` and this is causing the `starship_preexec` to fail with that error.

When installed via Nix, there is no automatic setup of starship, which means it doesn't install any of this stuff or take over your prompt. Therefore no error.

In any case, the right course of action here is to file an issue against [starship](https://github.com/starship/starship/) saying that the bash initialization fails when run with `set -u` and should be modified to ensure all variables should be referenced safely (e.g. `${STARSHIP_PREEXEC_READY-}`). In the meantime, you can just change your configuration to run `STARSHIP_PREEXEC_READY=false` prior to the starship initialization.

**Comment by Icy-Thought** [comment]
Resolved in Starship v0.58.0. Appreciate the effort of whomever took the time to implement the solution for this problem!
